### PR TITLE
Increase flexibility of TableBordersLayout

### LINKED ISSUE #957: More control over table borders
**Please explain your intent**
The Table feature is fantastic, but I find that I want more control over borders that is provided by TableBorderLayout.  I would really like to be able to control exactly which borders are turned on, and ideally also their thickness, independently.

**Describe the solution you'd like**
I was thinking that perhaps:

1. `TableBorderLayout` could become a regular class instead of an enum, so that a user could create instances of it
2. The existing values (e.g. `HORIZONTAL_LINES`) could still be class variables so that existing usage doesn't break
3. The class would hold on to a function/lamba that is called for every border segment for a table, and would return perhaps a thickness (0=off?) and a color (so table borders don't have to be black).

For (3) I'm imagining the function looking something like:
```python
def should_render(rownum: int, colnum: int, is_header: bool, is_footer:bool, side: ?) -> ...?
```
**Additional context**
I'm actually trying to port some nasty Rmarkdown -> latex -> PDF code, and users have gotten used to some latex table styling, which I can't reproduce.  One example is a table with i) thick borders above/below the table, ii) thin borders under the headerline, iii) no other borders.

### PR DESCRIPTION
**Pull request by TedBrookings**
Add capability to precisely control thickness, color, and dash of table cell borders.

`TableBordersLayout` is changed from an enum to a more complex class with helper style classes that allow more complex custom style setting. The old enum is maintained as static members of the new class, so the previous invocation style works the same, with the added option to specify cell border styles by passing a custom function.
Fixes #957

**Checklist**:
- [ ] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [ ] This PR is ready to be merged 

- [x] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

I'm making a draft PR. There are many unit tests that fail. The reason is partly down to strategy, and rather than plowing ahead and changing all the reference PDFs for the tests, I wanted to first get agreement that the overall approach I've used is good.
The tests fail for one of two reasons
1) The reference implementation draws two identical lines where mine draws one. This is easily fixable by just double-drawing the lines, and I've started doing that, but wanted agreement on this decision.
2) More critically, since the new class manages color, thickness, and dash qualities, I didn't want to managing setting and restoring all of these, so instead I wrap the drawing commands inside a local context (`q [set thickness/color/dash] [draw element] Q`). The result is visually identical and generally more concise, but it (correctly) fails `assert_pdf_equal`. My preferred solution would be to just update the test PDFs, but if this is not desired, it would be possible (if tedious) to adopt a set and reset approach that should yield identical outputs.

It also occurred to me that you might not like the class structure I created, so again I'd rather fix that and deal with the implications now, rather than plow ahead.

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by TedBrookings** [comment]
@Lucas-C @gmischler I've created a draft pull request in response to the issue. If you agree with the direction I've gone, it should be fairly quick to fix up remaining issues. Obviously I'm open to different direction if you see things you don't like.

**Comment by gmischler** [comment]
An interesting concept!
Although, looking at your documentation example and the test, creating a `TableBordersLayout()` seems to be quite a convoluted procedure. Having to provide several complicated lambdas in a constructor call is certainly not what I'd consider an intuitive API.
We try to keep fpdf2 straightforward and simple to use. The general idea of having a seperate class for managing borders and cell background may be helpful, but I think it needs to become much simpler and easier to understand how its rules are defined.

And a heads-up: I'm currently working on turning the table cells into text regions (see #339). To do that I'll have to rearrange quite a few parts of the table code, delegating tasks away from tha main `Table()` class. I hope this will simplify the overall architecture a bit. The borders code should not be directly affected by those changes, but I may still have to move it around.

**Comment by TedBrookings** [comment]
I understand the desire for simplicity.

To be fair, you only supply one lambda, and generally it would be quite simple (e.g. see the re-implementation of the existing styles). The demo was complicated to show all the things that could be done, but that could obviously be scaled back to provide a simple introduction, e.g. similar to MINIMAL but the line under headings is thick). The test needs to have every possible combination just in case a corner case breaks one of them.

**Comment by gmischler** [comment]
The specific number of lambdas is not the point.

Using query functions that the table elements can use to learn about their formatting is a useful concept. But that is an internal implementation detail, and not necessarily a good aproach for a public API. 
What you're doing here is essentially tell the user: "Just implement the formatting yourself!"
That's hardly more convenient than Subclassing `Table()` and replacing the border drawing method.

Rather than in 5 abstract lambda arguments, users may rather think in terms like "outer left border", "border below last heading", "bottom row of cells" etc. If you allow them to define rules in those terms and build your lambda out of them internally, then that would be a lot more helpful in terms of useability.

Forcing users to think in terms of internal implemention details is an issue I see with a lot of software on a daily basis. I really hope we can avoid falling into the same trap here.

**Comment by tfenne** [comment]
@gmischler weighing in here as I opened the original issue.  I wanted to share an example of a table style I'm trying to replicate.  It's toy data to illustrate, but hopefully it gets the point across:

image

Because (as I currently understand it) there's no way to embed tables in tables, I'm constructing this as a single table even though there are three sub-tables.  In this example I have:

- A major header row that labels each sub-table, with a single thin border below it
- A "minor" header row with column headers with a thick border below it
- A double-border above the summary/footer row
- Vertical divider lines between the "sub-tables" that start only at the "minor" header row

This is very similar to a style of table that is very frequently used in academic publications (at least in biology and I believe other sciences too) where people are used to seeing tables generated by latex.

I don't know the internals of the library nearly as well as you, but me sense is that if we only went with your suggestion (border styles settable for a specific categories of borders) I think either I wouldn't be able to do this _or_ there would be a be-wildering array of categories.

How would you feel if this was the underlying mechanism and the public API supported?:

1. The current method of supplying a named configuration of borders 
2. Something like you describe that provides a medium level of configurability by allowing the user to specify styles for a range of border categories
3. Approximately what Ted has implemented, as a last resort when people (like me) want to do things that can't be expressed by a more user-friendly API like (1) and (2)

**Comment by TedBrookings** [comment]
@gmischler I have been thinking about the best way to implement something along the lines of your suggestion. There are two issues that I see, and I see an easy work-around the first one, but I encountered some difficulties doing the second with a pure implementation.
1. It's entirely possible to define conflicting selectors with an API like the one you suggested. One thing to do would be to check for conflicts and throw an exception. Another would be to combine styles, treating later selectors as an override. This way you could e.g. specify gray color for all rows, but in a later call to the API make one of them thicker with a partially overlapping style that doesn't mention thickness.
2. I assume that its desirable to maintain existing `TableBorderStyle` values and the current method of specifying them, but to reimplement them in this new class. The problem being that some styles make use of the number of heading rows in the actual style definition. So for instance, I can't figure out a concise reimplementation of the `MINIMAL` table using an API similar to the one you propose, so that someone could create a minimal table like in this preexisting unit test [test_table_with_minimal_layout_and_multiple_headings](https://github.com/py-pdf/fpdf2/blob/874b9776b8dbb3fe0ac741be686f4d938cda3075/test/table/test_table.py#L572)

The best way forward that I can see to provide the API you like **_and_** reimplement the old styles is a hybrid approach. I could write a `TableBorderStyle` class with an API like the one you want, which stores data from the called APIs, and then implements its own `cell_style_getter` function based on those styles. So a user could if they chose just work with the API you wanted. This would probably be fine for one-off custom styles where they likely know the number of heading rows in advance.
Then the existing styles would be implemented by overwriting `cell_style_getter` with their own function that can make user of `num_header_rows`. This could be done in either of two ways:
1. Have an optional getter function (that defaults to `None`) as a member of the dataclass
2. Overwrite the class method for specific instances. e.g. something like `MINIMAL.cell_style_getter = some_lambda`. 

I think either approach would involve enough work that I'd like to get your thoughts before plowing ahead and coding.

**Comment by Lucas-C** [comment]
Hi @TedBrookings

I'm sorry that your last, detailed comment was left unanswered for a year 😢

Thank you very much for the effort you put in this PR 👍

Given that @gmischler did not answer your last comment,
I volunteer to take over in assisting you 🙂

But first, are you still willing to continue this work, almost a year later?
Or do you prefer that we simply close this PR?

I agree with the points you made in your last comment, about potential conflicting selectors, and the issue with reproducing the `MINIMAL` layout. Still, I'm not a big fan of the `cell_style_getter` lambda design

In `fpdf2` we added [TableCellFillMode](https://py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.TableCellFillMode) and the ability to override the [`should_fill_cell()` method](https://py-pdf.github.io/fpdf2/Tables.html#set-cells-background).
What do you think of this approach @TedBrookings?
Do you think that we could extend this approach to implement the remaining table styling you envisioned? (thickness, color, dashes...)

**Comment by TedBrookings** [comment]
Hi @Lucas-C I think we are interested in seeing this to completion one way or another. I just got back from international travel, but I should have time today to review my old work, `TableCellFillMode`, and issue 1992. After that I'll propose an approach that hopefully meets with your preferences and if you agree I can try and get this finished soon.

**Comment by Lucas-C** [comment]
Hi @TedBrookings! 🙂

What are is the status on your side regarding this PR?

Wish you a happy holiday season 🎄☃️ 𐂂

**Comment by TedBrookings** [comment]
Hi, I'm sorry for dropping out for a bit, things got very hectic with urgent client commitments.

I did look over the other work, and I think that unfortunately it's unlikely that we are going to agree on a path forward for this PR (and obviously this is your repo, so we respect your need to maintain your own coding standards!). The other PR added control over per-cell formatting, which my PR also did. So I would have to rebase my work to get back to the same place and use the new system. But that would still leave us at an impasse for how to define per-cell format.

If your objection is really just to _lambdas_, and you would be fine if the lambdas were replaced with ordinary functions in `def` statements, then that is a pretty trivial change, but my understanding is that the objection is to using functions whatsoever to define the format. Unfortunately I don't see how to do it otherwise.

The other suggestion involved explicitly passing lists of rows or columns for specific format effects, but in addition to being quite verbose, it then results in something that isn't a style, but a set of drawing instructions for each table (you have to know table dimensions exactly in advance). I also made a brief attempt at implementing them, but quickly found that coordinating between the different proposed format operations became much more complicated than simply specifying a per-cell format function. By contrast, e.g. [TableCellFillMode uses very simple functions to determine if a cell should be filled](https://github.com/py-pdf/fpdf2/blob/234cc331f5fd7b579d352852767ef13558374f6d/fpdf/enums.py#L409). Our original issue suggested allowing the definition of a new style that could be consistently applied to a different tables, and would hopefully be very succinct.

**Comment by Lucas-C** [comment]
> If your objection is really just to _lambdas_, and you would be fine if the lambdas were replaced with ordinary functions in def statements, then that is a pretty trivial change, but my understanding is that the objection is to using functions whatsoever to define the format.

Actually I think methods are fine, as we already used this approach for `TableCellFillMode`.
So adding methods to `TableBordersLayout` seems really OK to me.

I'm sorry for the inconsistent feedbacks we are giving you regarding this PR.
@gmischler was initially overseeing the work on this PR, and I have the greatest respect for his analysis.
But he has been less active on `fpdf2` for the pas few months, and I think that:
* this feature could benefit many users
* you produced some some substantial work in this PR, and were willing to listen to our feedbacks, and I think it is important to welcome quality contributions like yours
* even if your function-based approach do not cover all the cases, it could be quite user-friendly and consistent with the existing usages in `fpdf2`

So if you are still willing to finish the work started in this PR, I would be happy to merge a new version of this PR based on `TableBordersLayout` methods.

Maybe something akin to this:
```python
class TableBordersLayout:
    ...
    def border_style_for_cell(self, i, j, table):
        ...
```

**Comment by TedBrookings** [comment]
I think that seems practical. Just to be sure we're on the same page, you're thinking an abstract base class / protocol, or something similar, and then the user defines a new class with an actual implementation of `border_style_for_cell`?

In that case, I think the bulk of the actual work would just be rebasing to use the already merged support for defining Cell styles. I am pretty slammed for the rest of December, but if you're willing to wait until January, I think I could finish it off then.

**Comment by Lucas-C** [comment]
> Just to be sure we're on the same page, you're thinking an abstract base class / protocol, or something similar, and then the user defines a new class with an actual implementation of `border_style_for_cell`?

A base class yes, not necesserily an abstract one 🙂
We could provide a sane default class, as it's done with `TableCellFillMode`, but let users subclass this new class.

> In that case, I think the bulk of the actual work would just be rebasing to use the already merged support for defining Cell styles. I am pretty slammed for the rest of December, but if you're willing to wait until January, I think I could finish it off then.

That's perfectly fine, take your time 👍

**Comment by TedBrookings** [comment]
Hi @Lucas-C, I have made some progress on this, although there are two significant issues that I would like your input on:
1. Many tests fail now with differences in some `stream with MD5 hash` e.g. the only test to fail in `test_table.py` is `test/table/test_table.py::test_table_with_ttf_font_and_headings` has two of these hashes different. When I visually  inspect the expected and actual PDF, they are identical. Outside of `test_table.py`, there are many test cases that fail this way. I'm uncertain of the origin of this difference, so it's unclear the best way to proceed.
2. Existing table drawing methods only affect line color, and make the design decision to change color for a line by: changing to new color, drawing line, and restoring to old color. However my new `TableBorderStyle` can potentially adjust thickness and dash as well. To make tests pass, I have for now adopted the same strategy as the old one, just with more things being restored. There is an alternative which effectively creates a local context that prevents style changes from leaking. So it looks more like: start local context, alter style, draw line, end local context. Starting and ending local context are "q" and "Q" respectively, so they are very efficient. I would advocate for changing to this strategy, although it would require updating most of the old test PDFs. So it has a significant downside in tests.

**Comment by Lucas-C** [comment]
Hi @TedBrookings

Thank you for your continued work on this 🙂👍

I juste rebased your branch on this repo `master` branch.

> Many tests fail now with differences in some `stream with MD5 hash` e.g. the only test to fail in `test_table.py` is `test/table/test_table.py::test_table_with_ttf_font_and_headings` has two of these hashes different. When I visually inspect the expected and actual PDF, they are identical. Outside of `test_table.py`, there are many test cases that fail this way. I'm uncertain of the origin of this difference, so it's unclear the best way to proceed.

The `stream with MD5 hash` error is probably preceded by a warning,
explaining that this is a fallback check when `qpdf` is not available.
I would recommend you to install `qpdf` to have detailed diffs when a unit test fails.

If there are only invisible changes in the reference PDF files, then your changes are probably OK.

By the way, once you have committed some modified reference files on your branch, you can use `scripts/compare-changed-pdfs.py` to easily check for visual differences in them.

> Existing table drawing methods only affect line color, and make the design decision to change color for a line by: changing to new color, drawing line, and restoring to old color. However my new `TableBorderStyle` can potentially adjust thickness and dash as well. To make tests pass, I have for now adopted the same strategy as the old one, just with more things being restored. There is an alternative which effectively creates a local context that prevents style changes from leaking. So it looks more like: start local context, alter style, draw line, end local context. Starting and ending local context are "q" and "Q" respectively, so they are very efficient. I would advocate for changing to this strategy, although it would require updating most of the old test PDFs. So it has a significant downside in tests.

Your new `wrap_in_local_context()` function implementats context-wrapping,
which we already uses in the several places in `fpdf2` source code,
including in `FPDF._start_local_context()` & `FPDF._end_local_context()` methods.
We now have some minimal documention about this:
https://py-pdf.github.io/fpdf2/Internals.html#graphicsstatemixin
(it's not much, we should improve it progressively 😅)

It's fine to use it as you did in your new code in `fpdf/table.py`.
The best would maybe be to put this function in `fpdf/syntax.py,
and use it in every place that we currently _wrap_ stream content inbetween `q` & `Q` operators.
But we can do some refactoring later on 🙂

Also I see that in your branch you removed the `TableBordersLayout` enum from `fpdf.enums`.
This could be an issue, as this has been released a while ago, and many `fpdf2` users are probably importing this class from `fpdf.enums`.
We cannot simply remove it in the next release, as this would break a lot of codebases when end users upgrade to the latest `fpdf2` release.
Could you leave `TableBordersLayout` in `fpdf.enums`, even if it becomes a class, and extend its possibilities but also ensure that it can still be used in the same way as previously, please? 🙂

**Comment by TedBrookings** [comment]
@Lucas-C Hey, I should have enough time over the next week or so to bang this out and finally get it over the finish line.
I believe that I have made all the changes you requested.
1. I moved `TableBordersLayout` back to `enums.py`. This has the side effect of introducing some circular imports with `drawing.py`. I worked around this by importing inside a function, and then allowing this to pass `pylint` by disabling the lint for the two imports. I'm totally open to a different approach if you have something else in mind. **[EDIT: it looks like the github actions are different from the pre-commit hooks, and are not happy with this arrangement. I'll have to rethink]**
2. I moved the `wrap_in_local_context` to `syntax.py` as you requested.
3. I *do* have `qpdf` installed, and nevertheless many tests fail. I spot-checked them and they looked visually identical. To be systematic, I installed pypdfium2 and used it to render each page (at 4x scale) in the pdf and extract it to a numpy array, then asserted that the arrays were identical. All tests pass when I put this visual assert in as a backup to the main assert. I didn't commit this because it introduces a new dependency.

**Comment by Lucas-C** [comment]
> @Lucas-C Hey, I should have enough time over the next week or so to bang this out and finally get it over the finish line. I believe that I have made all the changes you requested.

Great! 👍

>     1. I moved `TableBordersLayout` back to `enums.py`. This has the side effect of introducing some circular imports with `drawing.py`. I worked around this by importing inside a function, and then allowing this to pass `pylint` by disabling the lint for the two imports. I'm totally open to a different approach if you have something else in mind. **[EDIT: it looks like the github actions are different from the pre-commit hooks, and are not happy with this arrangement. I'll have to rethink]**

Alright. Tell me if you need help at some point regarding this.

>     2. I moved the `wrap_in_local_context` to `syntax.py` as you requested.

👍

>     3. I _do_ have `qpdf` installed, and nevertheless many tests fail. I spot-checked them and they looked visually identical. To be systematic, I installed pypdfium2 and used it to render each page (at 4x scale) in the pdf and extract it to a numpy array, then asserted that the arrays were identical. All tests pass when I put this visual assert in as a backup to the main assert. I didn't commit this because it introduces a new dependency.

Alright, I'll check that if you are stuck on this.

**Comment by TedBrookings** [comment]
1. I was able to get the circular import work-around working with one more pylint disable.
2. I made a new commit with the new visually identical PDFs. I'm thinking that would make checks easier for you, although I'm happy to remove it if the reverse is the case.
3. I'm currently failing in `build / doc (pull_request)`. I'm not sure where to start on this one. I haven't been able to get `mkdocs` to work locally, I think in part because I'm not sure what all the required supporting packages are. At the moment I haven't put docstrings on every class and method, so I would want to clean that up before final PR. It's unclear from the error message if that's what's causing the crash

**Comment by TedBrookings** [comment]
@Lucas-C I've resolved all the GitHub actions problems. The `build / doc` crash seemingly was caused by a missing docstring. It should be ready for a detailed review.

File: fpdf/enums.py [hunk 1]
```diff
@@ -389,6 +367,547 @@ def __str__(self):
...
+    #         indexes into all cells including null ones. e.g. e.g. if there are two cells with
+    #         colspan = 3, then col_pos will be 0 or 3
+    #     num_heading_rows: the number of rows in the table heading
+    #     num_rows: the total number of rows in the table
+    #     num_col_idx: the number of non-null cells. e.g. if there are two cells with colspan = 3,
+    #         then num_col_idx = 2
+    #     num_col_pos: the full width of the table in physical cells. e.g. if there are two cells
+    #         with colspan = 3, then num_col_pos = 6
+    # Returns:
+    #     TableCellStyle for the given position in the table
```
Comment (Lucas-C) [inline fpdf/enums.py]: Could this be converted into a docstring of `cell_style_getter.()` please?

Also: shouldn't this method raise [NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError)?

**Review by Lucas-C — APPROVED** [review]
The code is really great: good job!

I made a single suggestion but overall this a very nice addition, I approve it 👍

There is a slight increase of 2 bytes per row for the PDF documents produced when using the new `TableBordersLayout` class, due to the usage of `wrap_in_local_context()`, but this is fine.

File: fpdf/enums.py (hunk 1 above)
Comment (TedBrookings) [reply fpdf/enums.py]: Fixed. The comment was a consequence of using the ellipsis, making the docstring invalid because ellipsis prevents any function body. I agree the docstring is much preferable, so I switched to that style and raised `NotImplementedError`.

**Comment by TedBrookings** [comment]
@Lucas-C I think I addressed your minor issue. You approved, but I'm not sure if you want to to the update or me, and I'm also not sure what the house style is (merge commit or rebase).

File: fpdf/enums.py (hunk 1 above)
Comment (Lucas-C) [reply fpdf/enums.py]: Perfect, thank you.

**Comment by Lucas-C** [comment]
For the record, this has been released in [version 2.8.3](https://github.com/py-pdf/fpdf2/releases/tag/2.8.3)
